<launch>

  <arg name="PROCESS" default="true" />
  <arg name="USE_PASSTHROUGH" default="false" />

  <node name="republish_tf_static"
        pkg="ros_objslampp_ycb_video" type="republish_tf_static_for_rosbag.py"
        clear_params="true">
  </node>

  <include file="$(find ros_objslampp_ycb_video)/launch/mask_rcnn_instance_segmentation.launch">
    <arg name="GUI" value="false" />
    <arg name="PROCESS" value="$(arg PROCESS)" />
    <arg name="USE_PASSTHROUGH" value="$(arg USE_PASSTHROUGH)" />
  </include>

  <include file="$(find ros_objslampp_ycb_video)/launch/singleview_3d_pose_estimation.launch">
    <arg name="FRAME_ID" value="map" />
    <arg name="STATIC_SCENE" value="true" />
    <arg name="OCCUPANCY" value="true" />
    <arg name="PROCESS" value="$(arg PROCESS)" />
  </include>

  <group ns="camera">
    <node name="render_voxel_grids"
          pkg="ros_objslampp_ycb_video" type="render_voxel_grids.py"
          clear_params="true" if="$(arg PROCESS)">
    </node>
    <node name="octomap_server"
          pkg="ros_objslampp_ycb_video" type="octomap_server"
          clear_params="true"
          output="screen" if="$(arg PROCESS)">
      <remap from="~render" to="render_voxel_grids/render" />
      <remap from="~input/camera_info" to="color/camera_info" />
      <remap from="~input/depth" to="aligned_depth_to_color/image_raw" />
      <remap from="~input/points" to="depth_registered/points" />
      <remap from="~input/label_ins" to="mask_rcnn_instance_segmentation/output/label_ins" />
      <remap from="~input/class" to="mask_rcnn_instance_segmentation/output/class" />
      <rosparam subst_value="true">
        frame_id: map
        resolution: 0.01
        ground_as_noentry: false
      </rosparam>
    </node>
    <node name="grids_to_mesh_markers"
          pkg="ros_objslampp_ycb_video" type="voxel_grids_to_mesh_markers.py"
          clear_params="true">
      <remap from="~input" to="octomap_server/output/grids" />
    </node>
    <node name="grids_to_markers"
          pkg="ros_objslampp_ycb_video" type="voxel_grids_to_markers.py"
          clear_params="true">
      <remap from="~input" to="octomap_server/output/grids" />
    </node>
    <node name="grids_noentry_to_markers"
          pkg="ros_objslampp_ycb_video" type="voxel_grids_to_markers.py"
          clear_params="true">
      <remap from="~input" to="octomap_server/output/grids_noentry" />
    </node>
    <node name="label_detected_to_rgb"
          pkg="ros_objslampp_ycb_video" type="label_image_decomposer.py"
          clear_params="true">
      <remap from="~input/label" to="mask_rcnn_instance_segmentation/output/label_ins" />
      <rosparam subst_value="true">
        only_label: true
        bg_label: -2
      </rosparam>
    </node>
    <node name="label_rendered_to_rgb"
          pkg="ros_objslampp_ycb_video" type="label_image_decomposer.py"
          clear_params="true">
      <remap from="~input/label" to="octomap_server/output/label_rendered" />
      <rosparam subst_value="true">
        only_label: true
        bg_label: -2
      </rosparam>
    </node>
    <node name="label_tracked_to_rgb"
          pkg="ros_objslampp_ycb_video" type="label_image_decomposer.py"
          clear_params="true">
      <remap from="~input/label" to="octomap_server/output/label_tracked" />
      <rosparam subst_value="true">
        only_label: true
        bg_label: -2
      </rosparam>
    </node>
    <group ns="octomap_server">
      <node name="draw_instance_segmentation"
            pkg="ros_objslampp_ycb_video" type="draw_instance_segmentation.py"
            clear_params="true">
        <remap from="~input/rgb" to="/camera/color/image_rect_color_passthrough/output" />
        <remap from="~input/label_ins" to="output/label_rendered" />
        <remap from="~input/class" to="output/class" />
        <rosparam subst_value="true">
          approximate_sync: true
        </rosparam>
      </node>
    </group>
  </group>

</launch>
